File: 2764-maximum-number-of-fish-in-a-grid/2764-maximum-number-of-fish-in-a-grid.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var findMaxFish = function (grid) {
    let visit = new Set();
    let max = 0;

    var dfs = function (start, end) {

        if (start < 0 ||
            end < 0 ||
            start >= grid.length ||
            end >= grid[0].length ||
            grid[start][end] == 0 ||
            visit.has(`${start},${end}`)) {
            return 0
        }

        visit.add(`${start},${end}`);


        let count = grid[start][end]

        for (let [x, y] of [[-1, 0], [0, -1], [1, 0], [0, 1]]) {
            let first = x + start;
            let second = y + end;


            count += dfs(first, second)

        }

        return count
    };

    for (let i = 0; i < grid.length; i++) {
        for (let j = 0; j < grid[0].length; j++) {
            if (grid[i][j] > 0 && !visit.has(`${i},${j}`)) {
                let all = dfs(i, j)
                max = Math.max(max, all);
            }
        }
    }
    return max
};

File: 2764-maximum-number-of-fish-in-a-grid/2764-maximum-number-of-fish-in-a-grid.py
class Solution:
    def findMaxFish(self, grid: List[List[int]]) -> int:
        visit = set()
        max_fish = 0

        def dfs(start, end):
            if (
                start < 0
                or end < 0
                or start >= len(grid)
                or end >= len(grid[0])
                or grid[start][end] == 0
                or (start, end) in visit
            ):
                return 0

            count = grid[start][end]
            visit.add((start, end))

            for x, y in [[-1, 0], [0, -1], [1, 0], [0, 1]]:
                first = x + start
                second = y + end

                count += dfs(first, second)

            return count

        for i in range(0, len(grid)):
            for j in range(0, len(grid[0])):
                if grid[i][j] > 0 and (i, j) not in visit:
                    allover = dfs(i, j)
                    max_fish = max(max_fish, allover)

        return max_fish
